src/V3Scope.cpp

```diff
@@ -260,7 +264,18 @@ class ScopeVisitor final : public VNVisitor {
     virtual void visit(AstVar* nodep) override {
         // Make new scope variable
         if (!nodep->user1p()) {
-            AstVarScope* const varscp = new AstVarScope(nodep->fileline(), m_scopep, nodep);
+            AstScope* scopep = m_scopep;
+            if (AstIfaceRefDType* const ifacerefp = VN_CAST(nodep->dtypep(), IfaceRefDType)) {
+                // Attach every non-virtual interface variable its inner scope
+                if (ifacerefp->cellp() && !ifacerefp->cellp()->isVirtIface())
+                    scopep = ifacerefp->cellp()->scopep();
+            }
+
+            AstVarScope* const varscp = new AstVarScope{nodep->fileline(), scopep, nodep};
+            if (m_aboveCellp && m_aboveCellp->isVirtIface()) {
+                // Prevent optimizations to this variable: may be cross-accessed
+                nodep->sigPublic(true);
```

Comment:
Is this unavoidable? Which optimization is this needed for?

Reply:
This could be avoided if we added special handling for scope overlaps with virtual interfaces in V3Dead and V3Const.

Without this deoptimization the provided regression test (`v8 = p8; v8.grant = 1; $display(v8.grant, p8.grant)`) shows the incorrect value for the actual interface (the value access to the virtual interface gets constified, and the assignment gets eliminated).  The variables are reported as unused, because each variable is in a different scope.